doc/scenario.md

```diff
@@ -1,72 +1,121 @@
 # Role Scenario
 
-The role uses a scenario to determine what to do.
-Any scenario consists of the previously defined steps.
-Using a scenario, you can:
-
-- run the specified steps in any order, but API must be followed
-  (with `cartridge_scenario` or `tasks_from`);
+Any Ansible role contains tasks.
+It's usually not possible to change set of tasks and tasks sequence. This is not good!
+Sometimes we need to run only some subset of tasks,
+execute tasks in a non-standard order or even add our own tasks.
+
+To do this, we grouped tasks in thematic sets, which we called [steps](/doc/steps.md). 
```

Comment:
To do this, we grouped tasks in thematic sets, which we called [steps](/doc/steps.md).

**Let's summarize:**
1. Scenario consists of steps;
2. Step consists of tasks.

## How to specify a scenario?

* Define a **scenario** (list of **steps**):
  ```yaml
  # playbook.yml
      vars:
        cartridge_scenario:
          - step_1
          - step_2
  ```

* Define a **scenario name** (identifier of some previously defined scenario):
  ```yaml
  # playbook.yml
      vars:
        cartridge_scenario_name: my_scenario
  # hosts.yml
  cartridge_custom_scenarios:
    my_scenario:
      - step_1
      - step_2
  ```

  You can use one of pre-defined scenarios (<link>)

## How to combine steps in a scenario?

Generally, you can place steps in any order, but [steps API](/doc/steps.md) must be followed.

What does it mean?

Take a look at the `deliver_package` and `update_package` steps.
They work together:

* `deliver_package` delivers package and set `delivered_package_path` variable;
* `update_package` installs package by `delivered_package_path`.

These steps use `delivered_package_path` variable to communicate with each other.
Each step described here <link> has some facts required for the step work and facts that the step produces.

## Is it possible to add custom steps?

Yes, using ... and ... variables (look at example)...

## Examples

### Writing a scenario using default steps

Imagine, we want to deliver the package on machines and install it.
Let's write a simple scenario:

```yaml
# playbook.yml
    vars:
      cartridge_scenario:
        - deliver_package
        - update_package
```

It looks that we will often use this scenario.
Save it to `cartridge_custom_scenarios` (e.g. in inventory) and then use by name:

```yaml
# hosts.yml
cartridge_custom_scenarios:
  deliver_and_update_package:
    - deliver_package
    - update_package
```

```yaml
# playbook.yml
    vars:
      cartridge_scenario_name: deliver_and_update_package
```

### Rewriting default steps

What if we want to download the package from some repository, but `deliver_package` step only copies the package from the local machine?

It's possible to rewrite default steps.
Create `custom_steps` directory and add `deliver_package.yml` step there:

```yaml
# ./custom_steps/deliver_package.yml

- when:
    - cartridge_package_path is not none
    - inventory_hostname in single_instances_for_each_machine
  tags: cartridge-instances
  block:
    - name: 'Download package'
      get_url:
        url: 'https://some-server/myapp-1.0.0-0.rpm'
        dest: '/tmp'
      register: downloaded_package

    - name: 'Set delivered package path'
      set_fact:
        delivered_package_path: '{{ downloaded_package.dest }}'
```

Import role and specify a path to `custom_steps` directory in `cartridge_custom_steps_dir` variable:

```yaml
# deploy_application.yml

- name: Deploy application
  hosts: all
  vars:
    cartridge_custom_steps_dir: "./custom_steps"
  roles:
    - tarantool.cartridge
```

Note, that:
* `single_instances_for_each_machine` <link> fact is used to download package once for each machine;
* `delivered_package_path` fact is set to respect API <link>;
* previously defined `deliver_and_update_package` scenario will use our custom `deliver_package`;

### Adding custom steps

...

## Importing steps from different directories

...

## Rolling update example

...

## Pre-defined scenarios

There are the role scenarios (scenario with `default` name is used by default):
- `default`:
  - [deliver_package](/doc/steps.md#deliver_package)